db/range_tombstone_fragmenter.cc

```diff
@@ -0,0 +1,256 @@
+  if (unfragmented_tombstones == nullptr) {
+    pos_ = tombstones_.end();
+    return;
+  }
+  bool is_sorted = true;
+  int num_tombstones = 0;
+  InternalKey pinned_last_start_key;
+  Slice last_start_key;
+  for (unfragmented_tombstones->SeekToFirst(); unfragmented_tombstones->Valid();
+       unfragmented_tombstones->Next(), num_tombstones++) {
+    if (num_tombstones > 0 &&
+        icmp_->Compare(last_start_key, unfragmented_tombstones->key()) > 0) {
+      is_sorted = false;
+      break;
+    }
+    if (unfragmented_tombstones->IsKeyPinned()) {
+      last_start_key = unfragmented_tombstones->key();
+    } else {
+      pinned_last_start_key.DecodeFrom(unfragmented_tombstones->key());
+      last_start_key = pinned_last_start_key.Encode();
+    }
+  }
+  if (is_sorted) {
+    FragmentTombstones(std::move(unfragmented_tombstones), snapshot);
+    return;
+  }
+
+  std::vector<std::string> keys, values;
+  keys.reserve(num_tombstones);
+  values.reserve(num_tombstones);
+  for (unfragmented_tombstones->SeekToFirst(); unfragmented_tombstones->Valid();
+       unfragmented_tombstones->Next()) {
+    keys.emplace_back(unfragmented_tombstones->key().data(),
+                      unfragmented_tombstones->key().size());
+    values.emplace_back(unfragmented_tombstones->value().data(),
+                        unfragmented_tombstones->value().size());
+  }
+  auto iter = std::unique_ptr<VectorIterator>(
+      new VectorIterator(std::move(keys), std::move(values), icmp_));
+  FragmentTombstones(std::move(iter), snapshot);
+}
+
+void FragmentedRangeTombstoneIterator::FragmentTombstones(
+    std::unique_ptr<InternalIterator> unfragmented_tombstones,
+    SequenceNumber snapshot) {
+  Slice cur_start_key(nullptr, 0);
+  auto cmp = ParsedInternalKeyComparator(icmp_);
+
+  // Stores the end keys and sequence numbers of range tombstones with a
+  // start key of cur_start_key. Provides an ordering by end key for use
+  // in flush_current_tombstones.
+  std::set<ParsedInternalKey, ParsedInternalKeyComparator> cur_end_keys(cmp);
+
+  // Given the next start key in unfragmented_tombstones,
+  // flush_current_tombstones writes every tombstone fragment that starts
+  // and ends with a key before next_start_key.
+  auto flush_current_tombstones = [&](const Slice& next_start_key) {
+    auto it = cur_end_keys.begin();
+    bool reached_next_start_key = false;
+    for (; it != cur_end_keys.end() && !reached_next_start_key; ++it) {
+      Slice cur_end_key = it->user_key;
+      if (icmp_->user_comparator()->Compare(cur_start_key, cur_end_key) == 0) {
+        // Empty tombstone.
+        continue;
+      }
+      if (icmp_->user_comparator()->Compare(next_start_key, cur_end_key) <= 0) {
+        reached_next_start_key = true;
+        cur_end_keys.erase(cur_end_keys.begin(), it);
+        cur_end_key = next_start_key;
+      }
+      SequenceNumber max_seqnum = 0;
+      for (auto flush_it = it; flush_it != cur_end_keys.end(); ++flush_it) {
```

Comment:
Think I finally understand this. One suggestion I have for readability is writing down the invariants and how they're maintained as much as possible. Invariants can also be "documented" by adding `assert`s, or even more complicated verification logic surrounded by `#ifndef NDEBUG`.

For example, I think here `[it, cur_end_keys.end())` all fully overlap the range `[cur_start_key, cur_end_key)`. And the reason for this (my understanding) is:
(1) `cur_end_keys` has been populated only with range tombstones that start before or at `cur_start_key`.
(2) `cur_end_key` is at most `*it`, and all later elements in `cur_end_keys` must be even larger than `*it`.

Reply:
Thanks for the suggestion; I'll add some invariant checks. And that invariant you gave is true, but not simple to verify because `cur_end_keys` doesn't store the start keys of the tombstones it contains.

Reply:
I only ended up adding one more assertion, but I also expanded on some comments (some of which help explain the example invariant you gave). I'll land this now, but let me know if there are other things that would benefit from more explanation/assertions, and I'll add them in the follow-up PR.